### ISSUE TITLE
Crash on Windows (was: High memory usage)

### ISSUE THREAD

**Issue by WhiteCat6142** [Bitmessage/PyBitmessage#964]
Bitmassage uses over 1GB memory.
It might be caused by a ton of threads for many connections.
I suggest using "gevent" to reduce the number.
Thank you.

**Comment by PeterSurda** [comment]
Duplicate of #870

**Comment by WhiteCat6142** [comment]
Sorry I didn't notice that.

**Comment by PeterSurda** [comment]
No problem, I hope it will be fixed in 0.6.3.

**Comment by chenshaoju** [comment]
Sorry for my English.

After upgrade to 0.6.2,The memory usage is too much.

On my Windows 7 SP1 x64 ,When Bitmessage running more than 8 hours, The bitmessage.exe use most 2GB memory.

I try set maxtotalconnections = 16 by edit key.dat file,but not work.
( https://twitter.com/PeterSurda/status/840029569235202048 )

This issue is not happen at 0.6.1.

![procxp](https://cloud.githubusercontent.com/assets/515763/23820150/30b860d4-064d-11e7-8f11-eb12cac58d57.png)

![vmmap](https://cloud.githubusercontent.com/assets/515763/23820151/34318330-064d-11e7-9bd8-c2cc304ee1cc.png)

If need more information please let'me know.

Thanks.

**Comment by PeterSurda** [comment]
Thank you, this is helpful. I think it's the threaded networking system causing this, it will be replaced in 0.6.3. I can send you a prerelease once it's done so that you can test it.

**Comment by WhiteCat6142** [comment]
it can be memory leak.
it uses 767MB for just 20 connections.

key.dat
>maxtotalconnections = 20

**Comment by PeterSurda** [comment]
I'm open to the idea that it's a memory leak but I'm also observing a lot of memory being used with 0.4.4 (for comparison), scaled down to the number of connections. The memory with 0.6.2 also seems to scale with the number of connections rather than uptime. I have nodes which have been running 0.6.2 since it was released. One has 300 connections and if you divide its memory allocation and compare it to another node with 0.4.4 which has 18, it seems to allocate the same amount of memory per connection.

The reasons why it looks like it was using less memory before is that the networking subsystem had bugs and deadlocks, the connections were short-lived and died frequently, so you couldn't have a large number of connections.

So in other words, it has required an excessive memory for a long time, but the bottleneck was in the network bugs. Now that that is fixed, the memory is the bottleneck.

**Comment by PeterSurda** [comment]
Please try the latest development snapshots (20170529) at https://bitmessage.org/download/snapshots/ and let me know if it's better. You can bump maxtotalconnections to 100 if you want.

**Comment by chenshaoju** [comment]
Test is running,I will tell you result ASAP.
And Thanks. :)

**Comment by chenshaoju** [comment]
After 9 hours test,The memory usage look's stable.

But this version (Bitmessagedev_x64_20170529.exe) bitmessage look's fully use one cpu core all the time.
![screenshot](https://cloud.githubusercontent.com/assets/515763/26565693/149d31ec-4520-11e7-82cf-3262f98f4120.png)

And this crashed one time,I set up a memory dump collection ,If crash again , I can send you the memory dump file.
```
错误应用程序名称: Bitmessagedev_x64_20170529.exe，版本: 0.0.0.0，时间戳: 0x00000000
错误模块名称: MSVCR90.dll，版本: 9.0.30729.6161，时间戳: 0x4dace4e7
异常代码: 0xc0000005
错误偏移量: 0x000000000001e2f5
错误进程 ID: 0x4e0c
错误应用程序启动时间: 0x01d2d884a24124ed
错误应用程序路径: D:\Software\BitMessager\Bitmessagedev_x64_20170529.exe
错误模块路径: C:\Windows\WinSxS\amd64_microsoft.vc90.crt_1fc8b3b9a1e18e3b_9.0.30729.6161_none_08e61857a83bc251\MSVCR90.dll
报告 ID: d5bb46d5-4492-11e7-88d9-0c8bfd2be3bd
```
```
错误存储段 ，类型 0
事件名称: APPCRASH
响应: 不可用
Cab ID: 0

问题签名: 
P1: Bitmessagedev_x64_20170529.exe
P2: 0.0.0.0
P3: 00000000
P4: MSVCR90.dll
P5: 9.0.30729.6161
P6: 4dace4e7
P7: c0000005
P8: 000000000001e2f5
P9: 
P10: 
```

**Comment by PeterSurda** [comment]
Thank you, I got other reports similar to yours. Unfortunately I don't know what to do with a windows coredump but maybe someone else can help. On Linux I haven't had a crash, but I occasionally get a "maximum recursion depth exceeded" error, maybe that's the same thing, I have a rough idea about what is causing it and hope to fix it within a couple of days. Once I can't see the errors anymore and I can't reproduce the crash on my Windows VMs anymore, I'll make a new snapshot.

**Comment by chenshaoju** [comment]
Thank you,Waiting new update. :100:

**Comment by Erkan-Yilmaz** [comment]
@PeterSurda 
in the current 0.6 branch I get such also:

_Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <_ctypes.DictRemover object at 0x7f2dcae4c570> ignored
Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <_ctypes.DictRemover object at 0x7f2dcae4ce50> ignored_

no crash yet, BM uses about 208 MB RAM, linux
(on request I can send more info)

**Comment by PeterSurda** [comment]
That's odd, I thought I fixed that recursion thing yesterday. Are you sure you're running the latest code? I don't see it anymore on any of my test linux systems and I haven't had a crash on a windows VM today either. One of my linux sytems has 132 connections and the process occupies slightly less than 1GB RAM.

**Comment by Erkan-Yilmaz** [comment]
> Are you sure you're running the latest code?

I am running [this branch](https://github.com/Bitmessage/PyBitmessage/tree/v0.6) with [this commit](https://github.com/Bitmessage/PyBitmessage/commit/4c17a1800648da0bcca5d1aca5302d361bc18e17) as last

**Comment by PeterSurda** [comment]
ok thanks I'll check it out.

**Comment by PeterSurda** [comment]
I see it again now.

**Comment by PeterSurda** [comment]
Can you try with the current code?

**Comment by chenshaoju** [comment]
I Tested Bitmessagedev_x64_20170609.exe ,Look's fine.
Memory usage 100MB~150MB.
No crash for last 24 hours.

**Comment by PeterSurda** [comment]
great :-). If you're having problems in the future, just try the latest dev snapshot, they are now generated automatically every day. I'll leave the ticket open because there are still some bugs I have to fix.

**Comment by chenshaoju** [comment]
It crash again :(
I guess need more details in logs,Like debug mode etc..?
```
描述
错误的应用程序路径:	D:\Software\BitMessager\Bitmessage.exe

问题签名
问题事件名称:	APPCRASH
应用程序名:	Bitmessage.exe
应用程序版本:	0.0.0.0
应用程序时间戳:	00000000
故障模块名称:	MSVCR90.dll
故障模块版本:	9.0.30729.6161
故障模块时间戳:	4dace4e7
异常代码:	c0000005
异常偏移:	000000000001e1ac
OS 版本:	6.1.7601.2.1.0.768.3
区域设置 ID:	2052
其他信息 1:	fffc
其他信息 2:	fffc335a66a691840c0514ee1f26bd5f
其他信息 3:	3379
其他信息 4:	3379ca903d96804751881fd642b35cb3
```

**Comment by Erkan-Yilmaz** [comment]
still seeing it in the 0.6 branch (A):

Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <_ctypes.DictRemover object at 0xaa9275c0> ignored
Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <_ctypes.DictRemover object at 0xaa927420> ignored
Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <_ctypes.DictRemover object at 0xaa9275d0> ignored

(A) with today's version, commit 189578c

**Comment by PeterSurda** [comment]
The recursion problem should now be handled (there will still be an error in the log, but it shouldn't crash anymore). Let me know if it still appears. For Windows, wait about 8.5 hours and try tomorrow's snapshot (20170707), they are now generated automatically.

**Comment by chenshaoju** [comment]
Thank you,I'm testing Bitmessagedev_x64_20170707.exe .

**Comment by PeterSurda** [comment]
Actually there are bugs in the latest code, you're better off waiting until tomorrow.

**Comment by chenshaoju** [comment]
OK,I'm wait,Thank you. :)

**Comment by chenshaoju** [comment]
The Bitmessagedev_x64_20170708 have a  strange issue:
After most 2 days,Bitmessage has disconnect form network ( red light ) 
I'm wating 1 hour,It still connecting.
I use Process Explorer to check Bitmessage TCP connect,It look's try connect to other client,but no one can connect.
![1](https://user-images.githubusercontent.com/515763/28000807-d0e2c286-655a-11e7-97fc-c3ed5607f516.png)

The Internet connection is working,And in my same network,Another Bitmessage still working fine. (0.6.1)

**Comment by PeterSurda** [comment]
Let me know if it gets better in tomorrow's snapshot.

**Comment by chenshaoju** [comment]
I tested Bitmessagedev_x64_20170711.exe ,After few hours,the Bitmessage has crashed.

![screenshot](https://user-images.githubusercontent.com/515763/28053162-9b510e06-6641-11e7-921f-835c85e41552.png)

**Comment by PeterSurda** [comment]
Thank you for the report, I'm working on it.

**Comment by chenshaoju** [comment]
I tested Bitmessagedev_x64_20170719,The disconnect still happen.
![1](https://user-images.githubusercontent.com/515763/28451813-2b134ff8-6e22-11e7-83a6-c20b21edbdb0.png)

**Comment by PeterSurda** [comment]
Try tomorrow's (July 22nd) snapshot, I'm still not sure if it fixes everything but maybe it will be better.

**Comment by chenshaoju** [comment]
I tested the Bitmessagedev_x64_20170723  version,The link look's keep connection ,but no sync and upload / download data.

I try send a message from another 0.6.1 client,It can't get the message.

If I restart the bitmessage,everything is working again.

![unnamed qq screenshot20170726092248](https://user-images.githubusercontent.com/515763/28601662-25b326a0-71ec-11e7-88bf-7f9781ba69dd.png)

**Comment by chenshaoju** [comment]
Update:
After test Bitmessagedev_x64_20171104.exe ,The Sync look's fine,No crash in 80 hours.

But memory usage still a bit high.

![screenshot](https://user-images.githubusercontent.com/515763/32497648-2d195f22-c408-11e7-9894-b67edc6dfbfb.png)

**Comment by PeterSurda** [comment]
I think most of the crashes have been solved and the rest of them I think I know how to fix. Regarding memory, the screenshot shows 44 connections and 374MB of RAM. The high "Virtual size" is probably due to how python works (on linux it also shows gigabytes of virtual memory for pybitmessage). It doesn't seem to have much practical meaning.

Currently, pybitmessage requires about 10MB per connections, which includes IO buffers and inventory tracking (queues are global so they don't require per-connection memory anymore). I don't think I can improve that much. The default connection limit of 200 should fit into 2GB or RAM.

My test instance has now been running for almost 2 weeks and peaked at 262 connections and about 2.6GB of RAM.

**Comment by PeterSurda** [comment]
That's probably the recursion bug which should be fixed soon. I'll let you know when the fix is in the snapshot.

**Comment by PeterSurda** [comment]
That's probably the recursion bug which should be fixed soon. I'll let you know when the fix is in the snapshot.

**Comment by PeterSurda** [comment]
I may have fixed the recursion bug (I'm not sure as I still don't have a reliable way to reproduce it). Try tomorrow's (20171123) snapshot.

**Comment by chenshaoju** [comment]
After 2 days,6 hours,The 20171124 version is crashed.

![image](https://user-images.githubusercontent.com/515763/33238641-a79d402e-d2cc-11e7-9340-40588b8a0478.png)

![image](https://user-images.githubusercontent.com/515763/33238651-da8ded12-d2cc-11e7-853f-08a22e084c1f.png)

![image](https://user-images.githubusercontent.com/515763/33238664-0085e722-d2cd-11e7-9849-8841323897dd.png)

**Comment by PeterSurda** [comment]
Thank you for the report. It looks like this is something Windows specific. I'll do more testing on Windows.

**Comment by PeterSurda** [comment]
Were you doing something with it when it crashed?

**Comment by chenshaoju** [comment]
The crash look's random,When I play games,watch youtube,or using ms word sometimes.

**Comment by PeterSurda** [comment]
Could it be that it has something to do with your nvidia card, e.g. a driver update? Try turning off OpenCL under the "Max acceptable difficulty" tab in the settings.

**Comment by PeterSurda** [comment]
I tested running PyBM while updating the nvidia driver (on Windows) and it seems to be fine.

**Comment by chenshaoju** [comment]
I tested 20171201 After 8 days running, everything look's fine for now.

![1](https://user-images.githubusercontent.com/515763/33803363-491d3632-ddc9-11e7-9d69-aa81af872036.png)

**Comment by PeterSurda** [comment]
Thank you for the report. I wasn't able to reproduce the crash at all on any systems. There may be / might have been something on your computer (e.g. a bug in the nvidia driver that was fixed in the meantime). I'll close it for now.

**Comment by chenshaoju** [comment]
after 10 days,it's crash again. :(

I'm not enable GPU accelerate or update drivers when this time.

![image](https://user-images.githubusercontent.com/515763/33894066-ff5baef4-df97-11e7-8a2b-29f0370d2ec9.png)

**Comment by PeterSurda** [comment]
how about the latest official release, 0.6.2, does that crash too?

**Comment by chenshaoju** [comment]
After update to Bitmessage_x86_0.6.3.2.exe, I moving the Bitmessage to a new server (Windows Server 2008 R2).

The server only has a built-in video card in CPU ( Intel(R) HD Graphics Family ).

But after 6 days, It's crashed.

![screenshot](https://user-images.githubusercontent.com/515763/36637803-0b4c7fe2-1a1f-11e8-9e30-8f0d12ea7e1b.png)

Log look does not have much info:
```
2018-02-19 21:43:24,529 - ERROR - 185.158.248.216:8444 error: 2, Server full, please try again later.
2018-02-19 21:43:24,660 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-19 21:43:24,914 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-19 21:43:36,694 - ERROR - C PoW test fail.
Traceback (most recent call last):
  File "proofofwork.py", line 256, in init
  File "proofofwork.py", line 106, in _doCPoW
ValueError: Procedure probably called with too many arguments (12 bytes in excess)
2018-02-19 21:43:39,671 - WARNING - No indicator plugin found
2018-02-19 21:43:39,671 - WARNING - No notification.message plugin found
2018-02-19 21:43:40,138 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-19 21:43:40,142 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-19 21:43:45,187 - ERROR - 185.158.248.216:8444 error: 2, Server full, please try again later.
2018-02-19 21:43:54,082 - ERROR - 85.25.152.9:8444 error: 2, Server full, please try again later.
2018-02-19 21:43:55,828 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-19 21:43:58,598 - ERROR - 93.211.161.55:8952 error: 2, Too many connections from your IP. Closing connection.
2018-02-19 23:20:00,190 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-20 02:04:09,714 - ERROR - 91.134.140.163:33574 error: 2, Too many connections from your IP. Closing connection.
2018-02-20 02:50:05,249 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-20 09:02:17,404 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-20 23:32:14,973 - ERROR - 85.114.135.102:33486 error: 2, Thank you for providing a listening node.
2018-02-21 00:18:21,407 - ERROR - 185.158.248.216:8444 error: 2, Server full, please try again later.
2018-02-21 00:29:38,483 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 00:41:24,918 - ERROR - 93.31.134.127:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 00:49:19,565 - ERROR - 93.31.134.127:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 01:15:43,921 - ERROR - 167.114.184.92:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 01:21:44,207 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-21 01:30:05,687 - ERROR - 95.85.40.163:8444 error: 2, Server full, please try again later.
2018-02-21 01:34:10,493 - ERROR - 93.31.134.127:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 01:49:09,490 - ERROR - 77.253.100.193:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 03:23:51,318 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-21 03:35:13,563 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 06:55:55,756 - ERROR - 86.217.109.186:39466 error: 2, Your time is too far in the past compared to mine. Closing connection.
2018-02-21 07:42:33,378 - ERROR - 86.217.109.186:46490 error: 2, Your time is too far in the past compared to mine. Closing connection.
2018-02-21 14:00:53,335 - ERROR - 144.76.78.214:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 18:27:03,924 - ERROR - 85.25.152.9:8444 error: 2, Server full, please try again later.
2018-02-21 18:30:44,598 - ERROR - 185.158.248.216:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 18:34:40,509 - ERROR - 185.158.248.216:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-21 18:38:29,170 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-21 22:00:16,635 - ERROR - 85.114.135.102:44955 error: 2, Thank you for providing a listening node.
2018-02-21 22:03:06,971 - ERROR - 85.25.152.9:46355 error: 2, Thank you for providing a listening node.
2018-02-21 23:26:45,194 - ERROR - 85.25.152.9:38332 error: 2, Thank you for providing a listening node.
2018-02-22 20:04:10,052 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-22 21:31:29,484 - ERROR - 85.114.135.102:45004 error: 2, Thank you for providing a listening node.
2018-02-22 23:05:00,082 - ERROR - 185.158.248.216:8444 error: 2, Server full, please try again later.
2018-02-22 23:09:02,663 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
2018-02-22 23:21:45,763 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-22 23:30:21,464 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-22 23:33:26,875 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-22 23:38:47,936 - ERROR - 144.76.78.214:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-23 00:09:17,947 - ERROR - 95.85.40.163:8444 error: 2, Server full, please try again later.
2018-02-23 00:55:11,812 - ERROR - 144.76.78.214:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-23 02:39:08,371 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-23 18:34:50,770 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-23 21:08:39,128 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-23 21:11:00,944 - ERROR - 144.76.78.214:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-23 21:28:32,421 - ERROR - 185.158.248.216:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-23 23:03:24,084 - ERROR - 85.114.135.102:58097 error: 2, Thank you for providing a listening node.
2018-02-24 00:16:03,865 - ERROR - 187.23.133.34:42608 error: 2, Your time is too far in the past compared to mine. Closing connection.
2018-02-24 01:42:25,951 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-24 02:06:02,674 - ERROR - Received command addr before connection was fully established, ignoring
2018-02-24 02:40:36,887 - ERROR - 77.253.100.193:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-24 02:51:12,997 - ERROR - 187.23.133.34:43956 error: 2, Your time is too far in the past compared to mine. Closing connection.
2018-02-24 05:39:40,401 - ERROR - 187.23.133.34:49402 error: 2, Your time is too far in the past compared to mine. Closing connection.
2018-02-24 09:04:07,759 - ERROR - 95.85.40.163:8444 error: 2, Server full, please try again later.
2018-02-24 11:18:49,664 - ERROR - 158.69.63.42:8080 error: 2, Too many connections from your IP. Closing connection.
2018-02-24 16:44:07,923 - ERROR - 95.85.40.163:8444 error: 2, Server full, please try again later.
2018-02-24 21:29:02,951 - ERROR - 85.114.135.102:43812 error: 2, Thank you for providing a listening node.
2018-02-24 22:16:29,141 - ERROR - 185.158.248.216:8444 error: 2, Server full, please try again later.
2018-02-24 22:36:12,157 - ERROR - 85.25.152.9:8444 error: 2, Server full, please try again later.
2018-02-24 22:36:36,937 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-24 22:45:41,007 - ERROR - 167.114.184.92:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-24 22:49:20,112 - ERROR - 158.69.63.42:8080 error: 2, Server full, please try again later.
2018-02-24 22:52:26,604 - ERROR - 93.31.134.127:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-24 22:56:27,999 - ERROR - 85.25.152.9:8444 error: 2, Server full, please try again later.
2018-02-24 22:56:32,108 - ERROR - 85.114.135.102:8444 error: 2, Server full, please try again later.
2018-02-24 23:01:25,194 - ERROR - 85.25.152.9:8444 error: 2, Server full, please try again later.
2018-02-24 23:39:59,862 - ERROR - 91.134.140.163:33574 error: 2, Too many connections from your IP. Closing connection.
2018-02-25 00:02:23,894 - ERROR - 144.76.78.214:8444 error: 2, Too many connections from your IP. Closing connection.
2018-02-25 00:19:35,276 - ERROR - 185.19.31.46:8080 error: 2, Server full, please try again later.
```

Crash report:
```
源
Bitmessage.exe

摘要
已停止工作

日期
‎2018/‎2/‎25 1:25

状态
已发送报告

描述
错误的应用程序路径:	D:\Software\Bitmessage\Bitmessage.exe

问题签名
问题事件名称:	APPCRASH
应用程序名:	Bitmessage.exe
应用程序版本:	0.0.0.0
应用程序时间戳:	5a2e9fe6
故障模块名称:	MSVCR90.dll
故障模块版本:	9.0.30729.4974
故障模块时间戳:	4b7a226f
异常代码:	c0000005
异常偏移:	0002463e
OS 版本:	6.1.7601.2.1.0.272.7
区域设置 ID:	2052
其他信息 1:	0a9e
其他信息 2:	0a9e372d3b4ad19135b953a78882e789
其他信息 3:	0a9e
其他信息 4:	0a9e372d3b4ad19135b953a78882e789

关于该问题的额外信息
存储段 ID:	2857640699
```

**Comment by PeterSurda** [comment]
Thank you for the info, @chenshaoju . Maybe it has to do with Chinese locale. I will try to get a W2008 with Chinese locale to test.

**Comment by PeterSurda** [comment]
It looks like the MSVCR90.DLL is not the latest one. I'll upgrade and rebuild.
